## **[Objectif: Maîtriser SQL]()** 
#### Le projet de cette semaine porte sur le SQL et la manipulation des bases des données relationnelles
#### Date de soumission: jeudi 15 juin 2023 à 22 H

##### Nous travaillerons dans cette partie avec une base des données du championnat de baskett ball américain (NBA)
##### Cette base est au format compressé et disponible publiquement dans kaggle et  accessible à l'adresse:
- [baskettball_data (url)](https://drive.google.com/file/d/17T1SQdChPn6JTpdLnd8VGWSOVGeWCL6W/view?usp=sharing)
- Vous devez obtenir une version décompresser directement dans votre environement de travail ou dans votre disque local

##### `Importer les librairies necéssaires pour travailler avec une application de gestion SQL avec python (sqlite3)`

In [1]:
import sqlite3
import pandas as pd

In [12]:
%load_ext sql
%sql sqlite:////Users/DDD/Downloads/data/data.sqlite

The sql extension is already loaded. To reload it, use:
  %reload_ext sql


**Tâche 1.1:** Initialiser sqlite3 et accéder à la base des données

In [6]:
## Vous devez modifier la séquence suivante (Users/DDD/Downloads/basketball.sqlite/basketball.sqlite) 
# en fonction de la localisation du fichier dans votre ordinateur
%load_ext sql
%sql sqlite:////Users/DDD/Downloads/basketball.sqlite/basketball.sqlite 

The sql extension is already loaded. To reload it, use:
  %reload_ext sql


**Tâche 1.2:** Faire votre première réquête dans la base en affichant son contenu
- [SQL: exposé de compréhension](https://www.i3s.unice.fr/~edemaria/cours/c4.pdf)
- [SQL: bases et fonctions](https://fxjollois.github.io/cours-sql/)
- Pour communiquer avec sqlite, chaque cellule de codes de jupyter doit commencer par '%%sql'

##### Utiliser la fonction SELECT  pour avoir accés aux informations et tables de la base

In [12]:
%%sql

SELECT *
FROM sqlite_schema

 * sqlite:////Users/DDD/Downloads/basketball.sqlite/basketball.sqlite
Done.


[('table', 'Player', 'Player', 2, 'CREATE TABLE "Player" (\n"id" TEXT,\n  "full_name" TEXT,\n  "first_name" TEXT,\n  "last_name" TEXT,\n  "is_active" INTEGER\n)'),
 ('table', 'Team', 'Team', 4, 'CREATE TABLE "Team" (\n"id" TEXT,\n  "full_name" TEXT,\n  "abbreviation" TEXT,\n  "nickname" TEXT,\n  "city" TEXT,\n  "state" TEXT,\n  "year_founded" INTEGER\n)'),
 ('table', 'Team_Attributes', 'Team_Attributes', 5, 'CREATE TABLE "Team_Attributes" (\n"ID" TEXT,\n  "ABBREVIATION" TEXT,\n  "NICKNAME" TEXT,\n  "YEARFOUNDED" TEXT,\n  "CITY" TEXT,\n  "ARENA" TEXT,\n  " ... (66 characters truncated) ...  "HEADCOACH" TEXT,\n  "DLEAGUEAFFILIATION" TEXT,\n  "FACEBOOK_WEBSITE_LINK" TEXT,\n  "INSTAGRAM_WEBSITE_LINK" TEXT,\n  "TWITTER_WEBSITE_LINK" TEXT\n)'),
 ('table', 'Team_History', 'Team_History', 6, 'CREATE TABLE "Team_History" (\n"ID" TEXT,\n  "CITY" TEXT,\n  "NICKNAME" TEXT,\n  "YEARFOUNDED" TEXT,\n  "YEARACTIVETILL" TEXT\n)'),
 ('table', 'Player_Attributes', 'Player_Attributes', 15, 'CREATE TABLE "Player_Attributes" (\n"ID" TEXT,\n  "FIRST_NAME" TEXT,\n  "LAST_NAME" TEXT,\n  "DISPLAY_FIRST_LAST" TEXT,\n  "DISPLAY_LAST_COMMA_FIRST" ... (569 characters truncated) ... R" TEXT,\n  "DRAFT_ROUND" TEXT,\n  "DRAFT_NUMBER" TEXT,\n  "PTS" REAL,\n  "AST" REAL,\n  "REB" REAL,\n  "ALL_STAR_APPEARANCES" REAL,\n  "PIE" REAL\n)'),
 ('table', 'Game_Officials', 'Game_Officials', 8375, 'CREATE TABLE "Game_Officials" (\n"OFFICIAL_ID" TEXT,\n  "FIRST_NAME" TEXT,\n  "LAST_NAME" TEXT,\n  "JERSEY_NUM" TEXT,\n  "GAME_ID" TEXT\n)'),
 ('table', 'Game_Inactive_Players', 'Game_Inactive_Players', 8997, 'CREATE TABLE "Game_Inactive_Players" (\n"PLAYER_ID" TEXT,\n  "FIRST_NAME" TEXT,\n  "LAST_NAME" TEXT,\n  "JERSEY_NUM" TEXT,\n  "TEAM_ID" TEXT,\n  "TEAM_CITY" TEXT,\n  "TEAM_NAME" TEXT,\n  "TEAM_ABBREVIATION" TEXT,\n  "GAME_ID" TEXT\n)'),
 ('table', 'Team_Salary', 'Team_Salary', 13415, 'CREATE TABLE `Team_Salary` (\n  `nameTeam` TEXT,\n  `slugTeam` TEXT,\n  `urlTeamSalaryHoopsHype` TEXT,\n  `X2020-21` REAL,\n  `X2021-22` REAL,\n  `X2022-23` REAL,\n  `X2023-24` REAL,\n  `X2024-25` REAL,\n  `X2025-26` REAL\n)'),
 ('table', 'Player_Salary', 'Player_Salary', 23558, 'CREATE TABLE `Player_Salary` (\n  `slugSeason` TEXT,\n  `nameTeam` TEXT,\n  `namePlayer` TEXT,\n  `statusPlayer` TEXT,\n  `isFinalSeason` INTEGER,\n  ... (31 characters truncated) ... ster` INTEGER,\n  `isNonGuaranteed` INTEGER,\n  `isTeamOption` INTEGER,\n  `isPlayerOption` INTEGER,\n  `typeContractDetail` TEXT,\n  `value` REAL\n)'),
 ('table', 'Draft', 'Draft', 23559, 'CREATE TABLE `Draft` (\n  `yearDraft` REAL,\n  `numberPickOverall` REAL,\n  `numberRound` REAL,\n  `numberRoundPick` REAL,\n  `namePlayer` TEXT,\n  ` ... (133 characters truncated) ... EXT,\n  `cityTeam` TEXT,\n  `teamName` TEXT,\n  `PLAYER_PROFILE_FLAG` REAL,\n  `slugOrganizationTypeFrom` TEXT,\n  `locationOrganizationFrom` TEXT\n)'),
 ('table', 'Draft_Combine', 'Draft_Combine', 23562, 'CREATE TABLE `Draft_Combine` (\n  `yearCombine` INTEGER,\n  `idPlayer` REAL,\n  `nameFirst` TEXT,\n  `nameLast` TEXT,\n  `namePlayer` TEXT,\n  `slugP ... (3799 characters truncated) ... legePct` REAL,\n  `setOffDribBreakRightCollegeMade` REAL,\n  `setOffDribBreakRightCollegeAttempted` REAL,\n  `setOffDribBreakRightCollegePct` REAL\n)'),
 ('table', 'Player_Photos', 'Player_Photos', 23565, 'CREATE TABLE `Player_Photos` (\n  `isActive` INTEGER,\n  `isRookie` INTEGER,\n  `namePlayer` TEXT,\n  `idPlayer` REAL,\n  `countSeasons` REAL,\n  `ye ... (170 characters truncated) ... dshot` TEXT,\n  `urlPlayerActionPhoto` TEXT,\n  `hasHeadShot` INTEGER,\n  `hasThumbnail` INTEGER,\n  `hasAction` INTEGER,\n  `urlPlayerPhoto` TEXT\n)'),
 ('table', 'Player_Bios', 'Player_Bios', 22541, 'CREATE TABLE `Player_Bios` (\n  `namePlayerBREF` TEXT,\n  `urlPlayerBioBREF` TEXT,\n  `nameTable` TEXT,\n  `urlPlayerImageBREF` TEXT,\n  `slugPlayerB ... (752 characters truncated) ...  `dateDeath` REAL,\n  `high schools` TEXT,\n  `descriptionRelatives` TEXT,\n  `descriptionHOF` TEXT,\n  `playerNickn

**Task 1.3:** Selectionner toutes les tables de la base
- Compléter le code ci-dessous en utilisant la fonction 'WHERE' 

In [16]:
%%sql

SELECT  name 
FROM sqlite_schema
WHERE type = "table"

 * sqlite:////Users/DDD/Downloads/basketball.sqlite/basketball.sqlite
Done.


[('Player',),
 ('Team',),
 ('Team_Attributes',),
 ('Team_History',),
 ('Player_Attributes',),
 ('Game_Officials',),
 ('Game_Inactive_Players',),
 ('Team_Salary',),
 ('Player_Salary',),
 ('Draft',),
 ('Draft_Combine',),
 ('Player_Photos',),
 ('Player_Bios',),
 ('Game',),
 ('News',),
 ('News_Missing',)]

**Task 1.4:** Select all columns from the `Player` table and from `Player_salary`, limiting your results to the first five rows. Use two separate cells to execute your code
- [SQL: bases et fonctions](https://fxjollois.github.io/cours-sql/)

In [17]:
%%sql
SELECT *
FROM Player
lIMIT 5

 * sqlite:////Users/DDD/Downloads/basketball.sqlite/basketball.sqlite
Done.


[('76001', 'Alaa Abdelnaby', 'Alaa', 'Abdelnaby', 0),
 ('76002', 'Zaid Abdul-Aziz', 'Zaid', 'Abdul-Aziz', 0),
 ('76003', 'Kareem Abdul-Jabbar', 'Kareem', 'Abdul-Jabbar', 0),
 ('51', 'Mahmoud Abdul-Rauf', 'Mahmoud', 'Abdul-Rauf', 0),
 ('1505', 'Tariq Abdul-Wahad', 'Tariq', 'Abdul-Wahad', 0)]

In [29]:
%%sql
SELECT *
FROM Player_salary
lIMIT 5

 * sqlite:////Users/DDD/Downloads/basketball.sqlite/basketball.sqlite
Done.


[('2020-21', 'Atlanta Hawks', 'Bogdan Bogdanovic', 'current roster', 0, 0, 1, 0, 0, 0, 'Guaranteed', 18000000.0),
 ('2021-22', 'Atlanta Hawks', 'Bogdan Bogdanovic', 'current roster', 0, 0, 1, 0, 0, 0, 'Guaranteed', 18000000.0),
 ('2022-23', 'Atlanta Hawks', 'Bogdan Bogdanovic', 'current roster', 0, 0, 1, 0, 0, 0, 'Guaranteed', 18000000.0),
 ('2023-24', 'Atlanta Hawks', 'Bogdan Bogdanovic', 'current roster', 1, 0, 1, 0, 0, 1, 'Player Option', 18000000.0),
 ('2020-21', 'Atlanta Hawks', 'Brandon Goodwin', 'current roster', 0, 0, 1, 0, 0, 0, 'Guaranteed', 1701593.0)]

**Task 1.5:** How many observations are in the `Player` table? Use the `count` command to find out. You should find no more than 4501 observations.
- [SQL count](https://www.sqlitetutorial.net/sqlite-count-function/)

In [4]:
%%sql
SELECT count(*)
FROM Player

 * sqlite:////Users/DDD/Downloads/basketball.sqlite/basketball.sqlite
Done.


[(4501,)]

**Task 1.6:** Select all columns from the `Team` table, limiting your results to the first five rows. Use `SELECT` and `LIMIT` functions
- [SQL: bases et fonctions](https://fxjollois.github.io/cours-sql/)

In [5]:
%%sql
SELECT *
FROM Team
LIMIT 5

 * sqlite:////Users/DDD/Downloads/basketball.sqlite/basketball.sqlite
Done.


[('1610612737', 'Atlanta Hawks', 'ATL', 'Hawks', 'Atlanta', 'Atlanta', 1949),
 ('1610612738', 'Boston Celtics', 'BOS', 'Celtics', 'Boston', 'Massachusetts', 1946),
 ('1610612739', 'Cleveland Cavaliers', 'CLE', 'Cavaliers', 'Cleveland', 'Ohio', 1970),
 ('1610612740', 'New Orleans Pelicans', 'NOP', 'Pelicans', 'New Orleans', 'Louisiana', 2002),
 ('1610612741', 'Chicago Bulls', 'CHI', 'Bulls', 'Chicago', 'Illinois', 1966)]

**Task 1.7:** What cities are represented in the `Team` table? Use the `distinct` command to determine the unique values in the **`city** column.
- [SQL: bases et fonctions](https://fxjollois.github.io/cours-sql/)

In [23]:
%%sql
SELECT distinct(city)
FROM Team

 * sqlite:////Users/DDD/Downloads/basketball.sqlite/basketball.sqlite
Done.


[('Atlanta',),
 ('Boston',),
 ('Cleveland',),
 ('New Orleans',),
 ('Chicago',),
 ('Dallas',),
 ('Denver',),
 ('Golden State',),
 ('Houston',),
 ('Los Angeles',),
 ('Miami',),
 ('Milwaukee',),
 ('Minnesota',),
 ('Brooklyn',),
 ('New York',),
 ('Orlando',),
 ('Indiana',),
 ('Philadelphia',),
 ('Phoenix',),
 ('Portland',),
 ('Sacramento',),
 ('San Antonio',),
 ('Oklahoma City',),
 ('Toronto',),
 ('Utah',),
 ('Memphis',),
 ('Washington',),
 ('Detroit',),
 ('Charlotte',)]

**Task 1.8:** How many cities are there in `Team` table? Combine the `count` and `distinct` commands to calculate the number of unique values in **`Team`**.

In [26]:
%%sql
SELECT count(distinct(city))
FROM team


 * sqlite:////Users/DDD/Downloads/basketball.sqlite/basketball.sqlite
Done.


[(29,)]

**Task 1.9:** For our model, we'll focus on slugSeason `2020-21`. Select all columns that from `Player_Salary`, showing only rows where the **`slugSeason`** is `2020-21` and limiting your results to the first thre rows.

In [36]:
%%sql
SELECT *
FROM player_salary
WHERE slugSeason= "2020-21"
lIMIT 3

 * sqlite:////Users/DDD/Downloads/basketball.sqlite/basketball.sqlite
Done.


[('2020-21', 'Atlanta Hawks', 'Bogdan Bogdanovic', 'current roster', 0, 0, 1, 0, 0, 0, 'Guaranteed', 18000000.0),
 ('2020-21', 'Atlanta Hawks', 'Brandon Goodwin', 'current roster', 0, 0, 1, 0, 0, 0, 'Guaranteed', 1701593.0),
 ('2020-21', 'Atlanta Hawks', 'Bruno Fernando', 'current roster', 0, 0, 1, 0, 0, 0, 'Guaranteed', 1517981.0)]

In [37]:
%%sql
SELECT *
FROM Player_attributes
lIMIT 5

 * sqlite:////Users/DDD/Downloads/basketball.sqlite/basketball.sqlite
Done.


[('76001', 'Alaa', 'Abdelnaby', 'Alaa Abdelnaby', 'Abdelnaby, Alaa', 'A. Abdelnaby', 'alaa-abdelnaby', '1968-06-24T00:00:00', 'Duke', 'USA', 'Duke/USA', 82.0, 240.0, 4, '30', 'Forward', 'Inactive', 'N', '1610612757', 'Trail Blazers', 'POR', 'blazers', 'Portland', 'HISTADD_alaa_abdelnaby', '1990', '1994', 'N', 'Y', 'Y', '1990', '1', '25', 5.7, 0.3, 3.3, 0.0, None),
 ('76002', 'Zaid', 'Abdul-Aziz', 'Zaid Abdul-Aziz', 'Abdul-Aziz, Zaid', 'Z. Abdul-Aziz', 'zaid-abdul-aziz', '1946-04-07T00:00:00', 'Iowa State', 'USA', 'Iowa State/USA', 81.0, 235.0, 9, '54', 'Center', 'Inactive', 'N', '1610612745', 'Rockets', 'HOU', 'rockets', 'Houston', 'HISTADD_zaid_abdul-aziz', '1968', '1977', 'N', 'Y', 'Y', '1968', '1', '5', 9.0, 1.2, 8.0, 0.0, None),
 ('76003', 'Kareem', 'Abdul-Jabbar', 'Kareem Abdul-Jabbar', 'Abdul-Jabbar, Kareem', 'K. Abdul-Jabbar', 'kareem-abdul-jabbar', '1947-04-16T00:00:00', 'UCLA', 'USA', 'UCLA/USA', 86.0, 225.0, 19, '33', 'Center', 'Inactive', 'N', '1610612747', 'Lakers', 'LAL', 'lakers', 'Los Angeles', 'HISTADD_kareem_abdul-jabbar', '1969', '1988', 'N', 'Y', 'Y', '1969', '1', '1', 24.6, 3.6, 11.2, 18.0, None),
 ('51', 'Mahmoud', 'Abdul-Rauf', 'Mahmoud Abdul-Rauf', 'Abdul-Rauf, Mahmoud', 'M. Abdul-Rauf', 'mahmoud-abdul-rauf', '1969-03-09T00:00:00', 'Louisiana State', 'USA', 'Louisiana State/USA', 73.0, 162.0, 8, '1', 'Guard', 'Inactive', 'N', '1610612743', 'Nuggets', 'DEN', 'nuggets', 'Denver', 'mahmoud_abdul-rauf', '1990', '2000', 'N', 'Y', 'Y', '1990', '1', '3', 14.6, 3.5, 1.9, 0.0, None),
 ('1505', 'Tariq', 'Abdul-Wahad', 'Tariq Abdul-Wahad', 'Abdul-Wahad, Tariq', 'T. Abdul-Wahad', 'tariq-abdul-wahad', '1974-11-03T00:00:00', 'San Jose State', 'France', 'San Jose State/France', 78.0, 235.0, 6, '9', 'Forward-Guard', 'Inactive', 'N', '1610612758', 'Kings', 'SAC', 'kings', 'Sacramento', 'tariq_abdul-wahad', '1997', '2003', 'N', 'Y', 'Y', '1997', '1', '11', 7.8, 1.1, 3.3, 0.0, None)]

**Task 1.10:** How many unique salary values do you have in the table create above? 
Remember that this table concerns only season 2020-21. Yous should have no more than 523 individuals salary.

In [35]:
%%sql
SELECT count(*)
FROM player_salary
WHERE slugSeason= "2020-21"

 * sqlite:////Users/DDD/Downloads/basketball.sqlite/basketball.sqlite
Done.


[(523,)]

**Task 1.11:** Copy and rename table and column in tha table. Use `ALTER` and `RENAME` functions.
You should first alter the name of the table as `salary_full_name` then rename `namePlayer` column as `full_name`.
`full_name` is the name of players in table `Player`.

In [62]:
%%sql
ALTER TABLE Player_salary
RENAME TO salary_full_name

 * sqlite:////Users/DDD/Downloads/basketball.sqlite/basketball.sqlite
Done.


[]

In [63]:
%%sql
ALTER TABLE salary_full_name
RENAME COLUMN namePlayer TO full_name

 * sqlite:////Users/DDD/Downloads/basketball.sqlite/basketball.sqlite
Done.


[]

In [64]:
%%sql
SELECT *
FROM salary_full_name
lIMIT 5

 * sqlite:////Users/DDD/Downloads/basketball.sqlite/basketball.sqlite
Done.


[('2020-21', 'Atlanta Hawks', 'Bogdan Bogdanovic', 'current roster', 0, 0, 1, 0, 0, 0, 'Guaranteed', 18000000.0),
 ('2021-22', 'Atlanta Hawks', 'Bogdan Bogdanovic', 'current roster', 0, 0, 1, 0, 0, 0, 'Guaranteed', 18000000.0),
 ('2022-23', 'Atlanta Hawks', 'Bogdan Bogdanovic', 'current roster', 0, 0, 1, 0, 0, 0, 'Guaranteed', 18000000.0),
 ('2023-24', 'Atlanta Hawks', 'Bogdan Bogdanovic', 'current roster', 1, 0, 1, 0, 0, 1, 'Player Option', 18000000.0),
 ('2020-21', 'Atlanta Hawks', 'Brandon Goodwin', 'current roster', 0, 0, 1, 0, 0, 0, 'Guaranteed', 1701593.0)]

In [87]:
%%sql
SELECT *
FROM Player
lIMIT 5

 * sqlite:////Users/DDD/Downloads/basketball.sqlite/basketball.sqlite
Done.


[('76001', 'Alaa Abdelnaby', 'Alaa', 'Abdelnaby', 0),
 ('76002', 'Zaid Abdul-Aziz', 'Zaid', 'Abdul-Aziz', 0),
 ('76003', 'Kareem Abdul-Jabbar', 'Kareem', 'Abdul-Jabbar', 0),
 ('51', 'Mahmoud Abdul-Rauf', 'Mahmoud', 'Abdul-Rauf', 0),
 ('1505', 'Tariq Abdul-Wahad', 'Tariq', 'Abdul-Wahad', 0)]

In [67]:
%%sql
SELECT count(*)
FROM salary_full_name

 * sqlite:////Users/DDD/Downloads/basketball.sqlite/basketball.sqlite
Done.


[(1292,)]

**Task .1.12:** Use the `distinct` command to create a column with all unique full_name in the `salary_full_name` table (the one you just created). `JOIN` this column with all the columns from the `Player` table, showing only Season where slugSeason is **2020-21**  and limiting your results to the first five rows of the new table
- [JOIN in SQLite](https://www.sqlitetutorial.net/sqlite-join/)

In [84]:
%%sql
SELECT *
FROM salary_full_name AS s
JOIN Player AS t ON s.full_name = t.full_name
WHERE slugSeason= "2020-21"
LIMIT 5

 * sqlite:////Users/DDD/Downloads/basketball.sqlite/basketball.sqlite
Done.


[('2020-21', 'Atlanta Hawks', 'Bogdan Bogdanovic', 'current roster', 0, 0, 1, 0, 0, 0, 'Guaranteed', 18000000.0, '203992', 'Bogdan Bogdanovic', 'Bogdan', 'Bogdanovic', 1),
 ('2020-21', 'Atlanta Hawks', 'Brandon Goodwin', 'current roster', 0, 0, 1, 0, 0, 0, 'Guaranteed', 1701593.0, '1629164', 'Brandon Goodwin', 'Brandon', 'Goodwin', 1),
 ('2020-21', 'Atlanta Hawks', 'Bruno Fernando', 'current roster', 0, 0, 1, 0, 0, 0, 'Guaranteed', 1517981.0, '1628981', 'Bruno Fernando', 'Bruno', 'Fernando', 1),
 ('2020-21', 'Atlanta Hawks', 'Cam Reddish', 'current roster', 0, 0, 1, 0, 0, 0, 'Guaranteed', 4458000.0, '1629629', 'Cam Reddish', 'Cam', 'Reddish', 1),
 ('2020-21', 'Atlanta Hawks', 'Clint Capela', 'current roster', 0, 0, 1, 0, 0, 0, 'Guaranteed', 16000000.0, '203991', 'Clint Capela', 'Clint', 'Capela', 1)]

##### Now you have 2 columns with full_name: full_name and full_name_1. We should rename full_name_1 as name

In [88]:
%%sql
SELECT distinct(s.full_name) AS name,
              s.*
FROM salary_full_name AS s
JOIN Player AS t ON s.full_name = t.full_name
WHERE slugSeason= "2020-21"
LIMIT 5

 * sqlite:////Users/DDD/Downloads/basketball.sqlite/basketball.sqlite
Done.


[('Bogdan Bogdanovic', '2020-21', 'Atlanta Hawks', 'Bogdan Bogdanovic', 'current roster', 0, 0, 1, 0, 0, 0, 'Guaranteed', 18000000.0),
 ('Brandon Goodwin', '2020-21', 'Atlanta Hawks', 'Brandon Goodwin', 'current roster', 0, 0, 1, 0, 0, 0, 'Guaranteed', 1701593.0),
 ('Bruno Fernando', '2020-21', 'Atlanta Hawks', 'Bruno Fernando', 'current roster', 0, 0, 1, 0, 0, 0, 'Guaranteed', 1517981.0),
 ('Cam Reddish', '2020-21', 'Atlanta Hawks', 'Cam Reddish', 'current roster', 0, 0, 1, 0, 0, 0, 'Guaranteed', 4458000.0),
 ('Clint Capela', '2020-21', 'Atlanta Hawks', 'Clint Capela', 'current roster', 0, 0, 1, 0, 0, 0, 'Guaranteed', 16000000.0)]

In [85]:
%%sql
SELECT count(*) 
FROM salary_full_name AS s
JOIN Player AS t ON s.full_name = t.full_name
WHERE slugSeason= "2020-21"

 * sqlite:////Users/DDD/Downloads/basketball.sqlite/basketball.sqlite
Done.


[(419,)]

**Task 1.13:** Put your the SQL query you have used to combine tables into a string and assign it to the variable `query`.

In [94]:
query = """ SELECT distinct(s.full_name) AS name,
              s.*
FROM salary_full_name AS s
JOIN Player AS t ON s.full_name = t.full_name
WHERE slugSeason= "2020-21"

"""
print(query)

 SELECT distinct(s.full_name) AS name,
              s.*
FROM salary_full_name AS s
JOIN Player AS t ON s.full_name = t.full_name
WHERE slugSeason= "2020-21"




**Task 1.14:** Use the [`connect`](https://docs.python.org/3/library/sqlite3.html#sqlite3.connect) method from the sqlite3 library to connect to the database. Remember that you should specify the url of your database.
- [Connection to a SQL database using sqlite3</span></span>.](../%40textbook/10-databases-sql.ipynb#Using-pandas-with-SQL-Databases)

In [83]:
conn = sqlite3.connect("C:/Users/DDD/Downloads/basketball.sqlite/basketball.sqlite")

**Task 1.15:** Use the [`read_sql`](https://pandas.pydata.org/docs/reference/api/pandas.read_sql.html#pandas-read-sql) from the pandas library to create a DataFrame from your `query`. Be sure that the **`full_name`** is set as your index column. You should not limit your query. 

In [95]:
import pandas as pd

df = pd.read_sql(query, conn, index_col="full_name")

df.head()

name slugSeason       nameTeam  \
full_name                                                        
Bogdan Bogdanovic  Bogdan Bogdanovic    2020-21  Atlanta Hawks   
Brandon Goodwin      Brandon Goodwin    2020-21  Atlanta Hawks   
Bruno Fernando        Bruno Fernando    2020-21  Atlanta Hawks   
Cam Reddish              Cam Reddish    2020-21  Atlanta Hawks   
Clint Capela            Clint Capela    2020-21  Atlanta Hawks   

                     statusPlayer  isFinalSeason  isWaived  isOnRoster  \
full_name                                                                
Bogdan Bogdanovic  current roster              0         0           1   
Brandon Goodwin    current roster              0         0           1   
Bruno Fernando     current roster              0         0           1   
Cam Reddish        current roster              0         0           1   
Clint Capela       current roster              0         0           1   

                   isNonGuaranteed  isTeamOption  isPlayerOption  \
full_name                                                          
Bogdan Bogdanovic                0             0               0   
Brandon Goodwin                  0             0               0   
Bruno Fernando                   0             0               0   
Cam Reddish                      0             0               0   
Clint Capela                     0             0               0   

                  typeContractDetail       value  
full_name                                         
Bogdan Bogdanovic         Guaranteed  18000000.0  
Brandon Goodwin           Guaranteed   1701593.0  
Bruno Fernando            Guaranteed   1517981.0  
Cam Reddish               Guaranteed   4458000.0  
Clint Capela              Guaranteed  16000000.0

In [96]:
df.shape

(419, 12)

In [97]:
df["value"].describe()

count    4.190000e+02
mean     8.128092e+06
std      9.702137e+06
min      5.000000e+04
25%      1.620564e+06
50%      3.872215e+06
75%      1.086595e+07
max      4.300636e+07
Name: value, dtype: float64

### **CREATION BASE DE DONNEES SQL A PARTIR DES FEUILLES DE CALCUL EXCEL**

In [1]:
import sqlite3
import pandas as pd

In [8]:
## Plusiuers feuilles de calcul
con=sqlite3.connect('database_ESR.sqlite')
excel_file = 'C:/Users/DDD/Documents/Cameroon_lung_function/SQLITE_tables_ESR/ESR_ready_1.xlsx'
wb=pd.ExcelFile(excel_file)  # lecture des feuilles avec pandas
for sheet in wb.sheet_names:
        df=pd.read_excel(excel_file,sheet_name=sheet)
        df.to_sql(sheet,con, index=False,if_exists="replace")
con.commit()
con.close()

In [2]:
## Plusiuers feuilles de calcul
con=sqlite3.connect('database_TP_14_07_2023.sqlite')
excel_file = 'C:/Users/DDD/Documents/Cameroon_lung_function/Cameroon_lung_function.xlsx'
wb=pd.ExcelFile(excel_file)  # lecture des feuilles avec pandas
for sheet in wb.sheet_names:
        df=pd.read_excel(excel_file,sheet_name=sheet)
        df.to_sql(sheet,con, index=False,if_exists="replace")
con.commit()
con.close()

In [6]:
%load_ext sql
%sql sqlite:////Users/DDD/Downloads/database_ESR.sqlite

The sql extension is already loaded. To reload it, use:
  %reload_ext sql


In [7]:
%%sql

SELECT *
FROM sqlite_schema

 * sqlite:////Users/DDD/Downloads/database_ESR.sqlite
Done.


[('table', 'table_demographie', 'table_demographie', 2, 'CREATE TABLE "table_demographie" (\n"NUM_ORDRE" INTEGER,\n  "GROUPE" INTEGER,\n  "Zone_denombrement" REAL,\n  "Numero_menage" INTEGER,\n  "Numéro_suj ... (277 characters truncated) ... AL,\n  "I.4_niveau_education_père" REAL,\n  "I.5_niveau_education_mère" REAL,\n  "I.6_Profession_principale" TEXT,\n  "CATEGORIE_PROFESS_REC" REAL\n)'),
 ('table', 'table_demographie_clinique', 'table_demographie_clinique', 116, 'CREATE TABLE "table_demographie_clinique" (\n"NUM_ORDRE" INTEGER,\n  "I.6_Profession_principale" TEXT,\n  "CATEGORIE_PROFESS_REC" REAL,\n  "I.7_catég ... (698 characters truncated) ... " TEXT,\n  "II.1.5_crachez_vous_habituellement" REAL,\n  "II.1.6_crachez_vous_laplupart_desjours" REAL,\n  "II.1.7_crachez_vous_3mois_ouplus" REAL\n)'),
 ('table', 'table_clinique', 'table_clinique', 196, 'CREATE TABLE "table_clinique" (\n"NUM_ORDRE" INTEGER,\n  "II.1.8_depuisconbien_années_crachez_vous" REAL,\n  "BRONCHITE_CHRONIQUE_REC" INTEGER,\n  "T ... (15179 characters truncated) ... ed_pre" REAL,\n  "severity_OLD_pre" REAL,\n  "fev_percent_predicted_post" REAL,\n  "severity_COPD" REAL,\n  "filter_$" INTEGER,\n  "bmi_copd" REAL\n)')]

In [ ]:
## une feuille de calcul
db = sqlite3.connect('sqlite.db')
dfs = pd.read_excel(excel_file , sheet_name=None)
for table, df in dfs.items():
    df.to_sql(table, db)
    print(f'{df} inserted successfully')